design_pim/overview.rst

```diff
@@ -0,0 +1,322 @@
+
+Simple business logic
++++++++++++++++++++++
+
+As an example, we will create a module that will display in the console the current UI locale
+
+To begin here is the business logic of our component:
+
+.. code-block:: javascript
+    :linenos:
+
+    {
+        logLocale: function () {
+            console.log(UserContext.get('uiLocale'));
+        }
+    }
+
+
+Our newly created object has one method which log the current catalog locale using the UserContext module.
+Now that we have our business logic implemented, we need to tell the system that our module has a dependency to ``pim/user-context`` and that we are exposing this object.
+
+The real thing
+++++++++++++++
+
+Here is how to do it using RequireJS:
+
+.. code-block:: javascript
+    :linenos:
+
+    // src/Acme/Bundle/CustomBundle/Resources/public/js/locale_logger.js
+    define(
+        ['pim/user-context'],
+        function (UserContext) {
+            return {
+                logLocale: function () {
+                    console.log(UserContext.get('uiLocale'));
+                }
+            }
+        }
+    )
+
+
+The define method takes two arguments: the first one is an array of the dependencies we want to use (``pim/user-context`` in this case) and a callback that will be called every time our module will be requested. This callback will receive as argument the list dependencies we asked for. Not that you can name the arguments of the callback as you want.
+
+Register it
++++++++++++
+
+Now that we have this small module, we would like to use it somewhere else. To do so, we need to register it in the requirejs configuration file.
+
+.. code-block:: yaml
+
+    # src/Acme/Bundle/CustomBundle/Resources/config/requirejs.yml
+    config:
+        paths:
+            my_locale_logger: acmecustom/js/locale_logger
+
+
+We just declared a new module named ``my_locale_logger`` located in ``web/bundles/acmecustom/js/locale_logger.js``.
+Not that we don't need to specify the file extension and that the location is the public web folder location (moved during the ``bin/console assets:install`` command execution). Note that the ``my_locale_logger`` name is a free name and you can replace it with whatever name you want.
+
+Make it available
++++++++++++++++++
+
+To make your module available you need to clear your cache:
+
+    rm -rf ./var/cache/*
+
+dump the assets in the web folder
+
+    bin/console assets:install web
+
+You can also decide to symlink those assets instead of copying them:
+
+    bin/console assets:install --symlink web
+
+After that the last step is to build the bundle.js file
+
+    npm run webpack
+
+This command will compile and minify all the pim files into one unique file and dump it in the public folder. You can also use the watch command which will recompile this file each time you modify a registered module.
```

Comment:
It dumps a few bundles, one for library modules, vendor, webpack util stuff and then the main app code